core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -648,6 +647,32 @@ private[spark] class BlockManager(
   }
 
   /**
+   * Retrieve the given block if it exists, otherwise call the provided `makeIterator` method
+   * to compute the block, persist it, and return its values.
+   */
+  def getOrElseUpdate(
+      blockId: BlockId,
+      level: StorageLevel,
+      makeIterator: () => Iterator[Any]): BlockResult = {
+    // Initially we hold no locks on this block.
+    doPut(blockId, IteratorValues(makeIterator), level, downgradeToReadLock = true)
```

Comment:
does anyone currently check the return value of doPut()? What if just returned Option[BlockResult] and then you can skip some of the logic below.